polyfill/intersectionobserver-polyfill.js

```diff
@@ -0,0 +1,231 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+(function() {
+  // TODO: rootMargin are completely ignored for now
+
+  // Constructor
+  window.IntersectionObserver = function(callback, options) {
+    options = options || {};
+
+    if(!(callback instanceof Function)) {
+      throw('callback needs to be a function');
+    }
+
+    if(options.root && !(options.root instanceof HTMLElement)) {
+      throw('Target needs to be an HTMLelement');
+    }
+
+    this._callback = callback;
+    this._root = options.root || null;
+    this._rootMargin = options.rootMargin || [0, 0, 0, 0];
+    this._thresholds = options.threshold || 0;
+    this._init();
+  };
+
+  window.IntersectionObserver.prototype = {
+    //
+    // Public API
+    //
+    get root() {
+      return this._root || document;
+    },
+
+    get rootMargin() {
+      return '0';
+    },
+
+    get thresholds() {
+      // 0 means call callback on every change
+      // See note at http://rawgit.com/WICG/IntersectionObserver/master/index.html#intersection-observer-init
+      if(this._thresholds === 0) {
+        return 0;
+      }
+      if(this._thresholds instanceof Array) {
+        return this._thresholds;
+      }
+      return [this._thresholds];
+    },
+
+    observe: function(target) {
+      if(!(target instanceof HTMLElement)) {
+        throw('Target needs to be an HTMLelement');
+      }
+
+      var root = this.root;
+      var ancestor = target.parentNode;
+      while (ancestor != root) {
+        if (!ancestor) {
+          throw('Target must be descendant of root');
+        }
+        ancestor = ancestor.parentNode;
+      }
+
+      this._mutationObserver.observe(target, {
+        attributes: true,
+        childList: true,
+        characterData: true,
+        subtree: true
+      });
+      this._observationTargets.set(target, {});
+    },
+
+    unobserve: function(target) {
+      this._observationTargets.delete(target);
+    },
+
+    disconnect: function() {
+      this._observationTargets.clear();
+      this.root.removeEventListener('scroll', this._boundUpdate);
+      this._mutationObserver.disconnect();
+      this._descheduleCallback();
+    },
+
+    takeRecords: function() {
+      this._update();
+      this._descheduleCallback();
+      var copy = this._queue.slice();
+      this._queue = [];
+      return copy;
+    },
+
+    //
+    // Private API
+    //
+    _init: function() {
+      this._observationTargets = new Map();
+      this._boundUpdate = this._update.bind(this);
+      this.root.addEventListener('scroll', debounce(this._boundUpdate, 100));
+      this._mutationObserver = new MutationObserver(debounce(this._boundUpdate, 100));
+      this._queue = [];
+    },
+
+    _update: function() {
+      var rootRect = this._rootRect();
+      this._observationTargets.forEach(function(oldIntersectionEntry, target) {
+        var targetRect = augmentRect(target.getBoundingClientRect());
```

Comment:
Personally I'd say abstracting the getBoundingClientRect call as well is probably clearer

```
getBoundingRect(target)

function getBoundingRect(el) {
    var r = el.getBoundingClientRect();
    r.width = r.width || r.right - r.left;
    r.height = r.height || r.bottom - r.top;
    return r;
 }
```